File: backend-csharp/TheGameIsOn.API/Services/EnemySpawnerService.cs
using TheGameIsOn.API.Models;
using TheGameIsOn.API.Repositories;

namespace TheGameIsOn.API.Services;

public record SpawnContext
{
    public MapArea? Area { get; init; }
    public string Difficulty { get; init; } = "normal";
    public int EncounterCount { get; init; }
    public string? EventId { get; init; }
}

public interface IEnemySpawnerService
{
    Task<List<BattleEnemy>> SpawnEnemies(SpawnContext context);
}

/// <summary>
/// Governs which enemies appear at the start of a battle node.
/// </summary>
public class EnemySpawnerService : IEnemySpawnerService
{
    private const int DefaultMinEnemies = 1;
    private const int DefaultMaxEnemies = 3;
    private const double LevelPerEncounter = 0.5;
    private const double HardDifficultyMultiplier = 1.4;
    private const double RewardScalePerLevel = 0.2;

    private readonly IEnemyRepository _enemyRepo;
    private readonly IEventRepository _eventRepo;
    private readonly ICardEffectRepository _cardEffectRepo;

    public EnemySpawnerService(
        IEnemyRepository enemyRepo,
        IEventRepository eventRepo,
        ICardEffectRepository cardEffectRepo)
    {
        _enemyRepo = enemyRepo;
        _eventRepo = eventRepo;
        _cardEffectRepo = cardEffectRepo;
    }

    public async Task<List<BattleEnemy>> SpawnEnemies(SpawnContext context)
    {
        MonsterSpawnConfig? spawnCfg = null;
        if (context.EventId is not null)
        {
            var eventDef = await _eventRepo.GetByIdAsync(context.EventId);
            if (eventDef?.MonsterSpawning is not null)
                spawnCfg = eventDef.MonsterSpawning;
        }

        var minEnemies = spawnCfg?.CountMin ?? DefaultMinEnemies;
        var maxEnemies = spawnCfg?.CountMax ?? DefaultMaxEnemies;
        var diffMod = spawnCfg?.DifficultyModifier ?? 1.0;

        var all = await _enemyRepo.GetAllAsync();

        // Filter by area and level
        var filterAreas = spawnCfg?.PoolFilter?.Areas;
        var filterMinLvl = spawnCfg?.PoolFilter?.MinLevel;
        var filterMaxLvl = spawnCfg?.PoolFilter?.MaxLevel;

        var pool = all.Where(def =>
        {
            if (filterAreas is { Count: > 0 })
            {
                if (def.SpawnArea is not null)
                {
                    var defArea = Enum.TryParse<MapArea>(def.SpawnArea, true, out var parsed) ? parsed : (MapArea?)null;
                    if (defArea.HasValue && !filterAreas.Contains(defArea.Value)) return false;
                }
            }
            else if (context.Area.HasValue)
            {
                if (def.SpawnArea is not null)
                {
                    var defArea = Enum.TryParse<MapArea>(def.SpawnArea, true, out var parsed) ? parsed : (MapArea?)null;
                    if (defArea.HasValue && defArea.Value != context.Area.Value) return false;
                }
            }

            if (filterMinLvl.HasValue && def.Level < filterMinLvl.Value) return false;
            if (filterMaxLvl.HasValue && def.Level > filterMaxLvl.Value) return false;
            return true;
        }).ToList();

        if (pool.Count == 0)
        {
            return await BuildEnemies(all.Take(minEnemies).ToList(), context, diffMod);
        }

        // Roll each candidate against its spawnChance
        var diffMult = (context.Difficulty == "hard" ? HardDifficultyMultiplier : 1.0) * diffMod;
        var rolled = pool.Where(def => Random.Shared.NextDouble() < def.SpawnChance * diffMult).ToList();

        // Enforce min / max counts
        List<EnemyDefinition> selected;
        if (rolled.Count == 0)
        {
            selected = pool.OrderByDescending(d => d.SpawnChance).Take(minEnemies).ToList();
        }
        else if (rolled.Count > maxEnemies)
        {
            selected = rolled.OrderByDescending(d => d.SpawnChance).Take(maxEnemies).ToList();
        }
        else
        {
            selected = rolled;
        }

        return await BuildEnemies(selected, context, diffMod);
    }

    // -------------------------------------------------------------------------

    private async Task<List<BattleEnemy>> BuildEnemies(
        List<EnemyDefinition> definitions, SpawnContext context, double difficultyModifier = 1.0)
    {
        var results = new List<BattleEnemy>();

        foreach (var def in definitions)
        {
            var level = ComputeLevel(def, context);
            var rewards = ScaleRewards(def.Rewards, def.Level, level);

            var levelDelta = level - def.Level;
            var hpMult = (1 + levelDelta * 0.15) * difficultyModifier;
            var life = (int)Math.Round(def.BaseLife * hpMult);

            var attackSummaries = new List<AttackSummary>();
            foreach (var atk in def.Attacks)
            {
                var effect = await _cardEffectRepo.GetByIdAsync(atk.EffectId);
                attackSummaries.Add(new AttackSummary
                {
                    Name = atk.Name,
                    Description = effect?.Description ?? "",
                    Element = atk.Element?.ToString()?.ToLowerInvariant()
                });
            }

            var uid = $"{def.Id}-{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}-{Guid.NewGuid().ToString("N")[..4]}";

            results.Add(new BattleEnemy
            {
                Id = uid,
                DefinitionId = def.Id,
                Name = def.Name,
                Type = def.Type,
                Element = def.Element?.ToString()?.ToLowerInvariant(),
                Life = life,
                MaxLife = life,
                Shield = 0,
                Energy = def.BaseEnergy,
                MaxEnergy = def.BaseEnergy,
                Level = level,
                ExpReward = (int)Math.Round(def.ExpReward * (1 + levelDelta * RewardScalePerLevel)),
                Rewards = rewards,
                AttackSummaries = attackSummaries
            });
        }

        return results;
    }

    private int ComputeLevel(EnemyDefinition def, SpawnContext context)
    {
        var bonus = (int)Math.Floor(context.EncounterCount * LevelPerEncounter);
        var diffBonus = context.Difficulty == "hard" ? 1 : 0;
        return def.Level + bonus + diffBonus;
    }

    private static List<EnemyReward> ScaleRewards(
        List<EnemyReward> rewards, int baseLevel, int computedLevel)
    {
        var levelDelta = computedLevel - baseLevel;
        if (levelDelta == 0) return rewards;

        return rewards.Select(r =>
        {
            if (r.Type == "gold" || r.Type == "exp")
            {
                return r with { Value = (int)Math.Round(r.Value * (1 + levelDelta * RewardScalePerLevel)) };
            }
            return r;
        }).ToList();
    }
}

File: backend/src/services/enemy-spawner.service.ts
import { EnemyRepository } from '../repo/enemy-repo';
import { EventRepository } from '../repo/event-repo';
import { CardEffectRepository } from '../repo/card-effect-repo';
import { EnemyDefinition, EnemyReward } from '../models/enemy';
import { BattleEnemy } from '../models/battle-state';
import { MonsterSpawnConfig } from '../models/event-definition';
import { MapArea } from '../models/node-event';

export interface SpawnContext {
  /** Zone the current node belongs to. Filters eligible enemy pool. */
  area: MapArea | undefined;
  /** 'normal' | 'hard' — hard nodes apply an extra difficulty multiplier. */
  difficulty: 'normal' | 'hard';
  /** Total battles the player has completed so far in this session. */
  encounterCount: number;
  /** Optional event id — when provided, MonsterSpawnConfig from events.json overrides defaults. */
  eventId?: string;
}

/**
 * Governs which enemies appear at the start of a battle node.
 *
 * Spawning algorithm
 * ──────────────────
 * 1. Filter the enemy pool to those whose `spawnArea` matches the current
 *    node area (or is null/undefined — those spawn everywhere).
 * 2. Roll each candidate against its `spawnChance`, scaled by difficulty.
 * 3. Guarantee at least MIN_ENEMIES and cap at MAX_ENEMIES.
 * 4. Scale enemy level: base level from JSON + bonus from encounterCount.
 * 5. Scale rewards (gold, exp) proportionally to computed level.
 */
export class EnemySpawnerService {
  private readonly DEFAULT_MIN_ENEMIES = 1;
  private readonly DEFAULT_MAX_ENEMIES = 3;
  /** Each completed encounter adds this fraction to the level bonus. */
  private readonly LEVEL_PER_ENCOUNTER = 0.5;
  /** Hard-battle nodes multiply spawnChance by this factor. */
  private readonly HARD_DIFFICULTY_MULTIPLIER = 1.4;
  /** Multiplier applied to gold/exp rewards for each level above 1. */
  private readonly REWARD_SCALE_PER_LEVEL = 0.2;

  private enemyRepo = new EnemyRepository();
  private eventRepo = new EventRepository();
  private cardEffectRepo = new CardEffectRepository();

  /**
   * Build the list of `BattleEnemy` instances to place in a new BattleState.
   *
   * When `context.eventId` is set the spawn config (pool filter, count range,
   * difficulty modifier) is read from the matching EventDefinition in
   * events.json, overriding the built-in defaults.
   */
  async spawnEnemies(context: SpawnContext): Promise<BattleEnemy[]> {
    // Resolve MonsterSpawnConfig from events.json when an eventId is provided.
    let spawnCfg: MonsterSpawnConfig | null = null;
    if (context.eventId) {
      const eventDef = await this.eventRepo.getById(context.eventId);
      if (eventDef?.monsterSpawning) {
        spawnCfg = eventDef.monsterSpawning;
      }
    }

    const minEnemies = spawnCfg?.countMin ?? this.DEFAULT_MIN_ENEMIES;
    const maxEnemies = spawnCfg?.countMax ?? this.DEFAULT_MAX_ENEMIES;
    const diffMod    = spawnCfg?.difficultyModifier ?? 1;

    const all = await this.enemyRepo.getAll();

    // 1. Filter by area — use poolFilter.areas from event config if available,
    //    otherwise fall back to the node area.
    const filterAreas = spawnCfg?.poolFilter?.areas;
    const filterMinLvl = spawnCfg?.poolFilter?.minLevel;
    const filterMaxLvl = spawnCfg?.poolFilter?.maxLevel;

    let pool = all.filter(def => {
      // Area filter
      if (filterAreas && filterAreas.length > 0) {
        if (def.spawnArea && !filterAreas.includes(def.spawnArea as any)) return false;
      } else if (context.area) {
        if (def.spawnArea && def.spawnArea !== context.area) return false;
      }
      // Level filter
      if (filterMinLvl !== undefined && def.level < filterMinLvl) return false;
      if (filterMaxLvl !== undefined && def.level > filterMaxLvl) return false;
      return true;
    });

    if (pool.length === 0) {
      // Fallback: use the entire roster so a battle is never empty.
      return this.buildEnemies(all.slice(0, minEnemies), context, diffMod);
    }

    // 2. Roll each candidate against its spawnChance.
    const diffMult =
      (context.difficulty === 'hard' ? this.HARD_DIFFICULTY_MULTIPLIER : 1) * diffMod;

    const rolled = pool.filter(def => Math.random() < def.spawnChance * diffMult);

    // 3. Enforce min / max counts.
    let selected: EnemyDefinition[];
    if (rolled.length === 0) {
      const sorted = [...pool].sort((a, b) => b.spawnChance - a.spawnChance);
      selected = sorted.slice(0, minEnemies);
    } else if (rolled.length > maxEnemies) {
      const sorted = [...rolled].sort((a, b) => b.spawnChance - a.spawnChance);
      selected = sorted.slice(0, maxEnemies);
    } else {
      selected = rolled;
    }

    return this.buildEnemies(selected, context, diffMod);
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  private async buildEnemies(
    definitions: EnemyDefinition[],
    context: SpawnContext,
    difficultyModifier: number = 1,
  ): Promise<BattleEnemy[]> {
    return Promise.all(definitions.map(async def => {
      const level   = this.computeLevel(def, context);
      const rewards = this.scaleRewards(def.rewards, def.level, level);

      // Scale HP and energy linearly with level delta + difficulty modifier.
      const levelDelta  = level - def.level;
      const hpMult      = (1 + levelDelta * 0.15) * difficultyModifier;  // +15 % HP per extra level
      const life        = Math.round(def.baseLife   * hpMult);

      // Build attack summaries from definition + card-effects descriptions.
      const attackSummaries = await Promise.all(
        (def.attacks ?? []).map(async atk => {
          const effect = await this.cardEffectRepo.getById(atk.effectId);
          return {
            name: atk.name,
            description: effect?.description ?? '',
            element: atk.element,
          };
        })
      );

      return {
        id:           `${def.id}-${Date.now()}-${Math.random().toString(36).slice(2, 6)}`,
        definitionId: def.id,
        name:         def.name,
        type:         def.type,
        element:      def.element,
        life,
        maxLife:      life,
        shield:       0,
        energy:       def.baseEnergy,
        maxEnergy:    def.baseEnergy,
        level,
        expReward:    Math.round(def.expReward * (1 + levelDelta * this.REWARD_SCALE_PER_LEVEL)),
        rewards,
        attackSummaries,
      };
    }));
  }

  /**
   * Compute the runtime level for one enemy.
   *
   * level = max(def.level,  def.level + floor(encounterCount * LEVEL_PER_ENCOUNTER))
   *
   * Example: a level-1 wolf after 4 encounters → level 1 + floor(4*0.5) = level 3.
   */
  private computeLevel(def: EnemyDefinition, context: SpawnContext): number {
    const bonus = Math.floor(context.encounterCount * this.LEVEL_PER_ENCOUNTER);
    // Hard battles add one extra level.
    const diffBonus = context.difficulty === 'hard' ? 1 : 0;
    return def.level + bonus + diffBonus;
  }

  /**
   * Scale gold and exp rewards based on the computed level vs. the base level.
   * card-draw rewards are kept as-is (tier is fixed by design).
   */
  private scaleRewards(
    rewards: EnemyReward[],
    baseLevel: number,
    computedLevel: number
  ): EnemyReward[] {
    const levelDelta = computedLevel - baseLevel;
    if (levelDelta === 0) return rewards;

    return rewards.map(r => {
      if (r.type === 'gold' || r.type === 'exp') {
        return {
          ...r,
          value: Math.round(r.value * (1 + levelDelta * this.REWARD_SCALE_PER_LEVEL)),
        };
      }
      return r;
    });
  }
}
